### DISCUSSION TITLE
Feature: EdgeDB datasource

### DISCUSSION THREAD

**Discussion by techninja1008** [Budibase/budibase#4860] [Ideas]
**Describe the feature request**

Add a datasource for [EdgeDB](https://www.edgedb.com/) to budibase, ideally with full support for links, property cardinality and a reasonable mapping of EdgeDB's type system.

**Screenshots** N/A

**Implementation**

A significantly simpler version of the integration could be done as an MVP to investigate how well it would work and how best to move forwards.

I'll willing to attempt a contribution of this feature if given some direction on what areas of the codebase would need touching and any special considerations to be made.

**Comment by mike12345567** [comment]
Hi @techninja1008 - thanks for the suggestion, I've converted this to an idea discussion so that others can discuss it as well.

In terms of direction for contributing a feature like this, we actually have some documentation around writing a new external data source integration: https://docs.budibase.com/references-1/contributing-to-budibase/contribute-data-connectors

**Reply by techninja1008** [reply]
Thanks for the direction - I've read through that guide and had a look at the source a bit more. What I'm struggling with is how to properly integrate EdgeDB's links - from the perspective of Budibase, these should automatically appear as relationships on the correct tables, but I'm not sure if it's possible for buildSchema on the integration to specify that without it breaking elsewhere. Also, none of the logic surrounding foreign keys is useful for EdgeDB, nor any many-to-many logic, as it's handled directly by the database.

Is there any way of specifying that an integration is able to handle all the relationship logic itself?

**Reply by mike12345567** [reply]
Hi @techninja1008 - currently Budibase has a two concepts of datasource:
1. A standard datasource, this allows connecting to a database and executing queries, e.g. MongoDB, CouchDB, DynamoDB etc
2. A datasource "plus", this allows connecting to an external database and building a schema about it, however currently we have only built up support for SQL based datasource plus entities. Datasource plus databases appear as separate tables within Budibase and are supposed to have near identical functionality to our internal database.

What you are talking about is the datasource plus concept, right now we really only support SQL type databases for this as there is a lot of complexity around all of the functionality that we support here.

At some point in the future we will likely start to support NoSQL and other database types in a datasource plus fashion, as part of this we will start to break up our external tables/rows controllers and plus integrations for databases like EdgeDB will become possible.

For the time being, its likely only a standard datasource could be implemented, allowing execution of queries on the database.

If you're interested the following files are involved in datasource plus entities:
1. [external table controller](https://github.com/Budibase/budibase/blob/develop/packages/server/src/api/controllers/table/external.js) - this allows creating, updating deleting external tables.
2. [external row controller](https://github.com/Budibase/budibase/blob/develop/packages/server/src/api/controllers/row/external.js) - handles all the components of the rows API that need to be supported.
3. [External request generation](https://github.com/Budibase/budibase/blob/develop/packages/server/src/api/controllers/row/ExternalRequest.ts) - right now this is very specific to SQL, this generates a JSON structure which the underlying datasource will eventually convert into queries.
4. [SQL data queries](https://github.com/Budibase/budibase/blob/develop/packages/server/src/integrations/base/sql.ts) - this converts the JSON generated from the rows API into queries for the SQL datasource.
5. [SQL table queries](https://github.com/Budibase/budibase/blob/develop/packages/server/src/integrations/base/sqlTable.ts) - this converts the JSON generated from the tables API into queries for the SQL datasource.

As you can see the datasource plus model is quite involved and currently geared towards supported SQL datasources, as a minimum, the SQL data query layer would need to be replicated for EdgeDB, with a mechanism to convert the JSON structure into EdgeDB queries automatically - as well as some work to change how the external request generation thinks about relationships.

**Reply by techninja1008** [reply]
Thanks Michael - this was the general gist I was getting from reading through the codebase. I'll probably start with a standard datasource then that supports EdgeQL queries (EdgeDB's query language) and then investigate at some point to see how well adding datasource plus support would be. I know it would definitely fit quite well in terms of the model that the user sees - EdgeDB actually matches Budibase's data model very closely with a few exceptions.
